### PR TITLE
[SPARK-8583] [SPARK-5482] [BUILD] Refactor python/run-tests to integrate with dev/run-tests module system

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch refactors the `python/run-tests` script:
- It's now written in Python instead of Bash.
- The descriptions of the tests to run are now stored in `dev/run-tests`'s modules.  This allows the pull request builder to skip Python tests suites that were not affected by the pull request's changes.  For example, we can now skip the PySpark Streaming test cases when only SQL files are changed.
- `python/run-tests` now supports command-line flags to make it easier to run individual test suites (this addresses SPARK-5482):
  
  ```
  Usage: run-tests [options]
  
  Options:
  -h, --help            show this help message and exit
  --python-executables=PYTHON_EXECUTABLES
                        A comma-separated list of Python executables to test
                        against (default: python2.6,python3.4,pypy)
  --modules=MODULES     A comma-separated list of Python modules to test
                        (default: pyspark-core,pyspark-ml,pyspark-mllib
                        ,pyspark-sql,pyspark-streaming)
  ```
- `dev/run-tests` has been split into multiple files: the module definitions and test utility functions are now stored inside of a `dev/sparktestsupport` Python module, allowing them to be re-used from the Python test runner script.

### CODE REVIEW

File: python/run-tests [hunk 1]
```diff
@@ -18,165 +18,9 @@
...
-cd "$FWDIR/python"
-
-FAILED=0
-LOG_FILE=unit-tests.log
-START=$(date +"%s")
-
-rm -f $LOG_FILE
-
-# Remove the metastore and warehouse directory created by the HiveContext tests in Spark SQL
-rm -rf metastore warehouse
```
Comment (JoshRosen) [inline python/run-tests]: @ahirreddy, I think that you originally added these lines.  Do we still need them?  AFAIK the pull request builder's `git clean -fdx` should take care of this for us, but was there another reason why we need this?

**Comment by JoshRosen** [comment]
Haha, this is now failing the tests for the test runner script:

```python
**********************************************************************
File "./dev/run-tests.py", line 41, in __main__.determine_modules_for_files
Failed example:
    sorted(x.name for x in determine_modules_for_files(["python/pyspark/a.py", "sql/test/foo"]))
Expected:
    ['pyspark', 'sql']
Got:
    ['pyspark-core', 'sql']
**********************************************************************
File "./dev/run-tests.py", line 93, in __main__.determine_modules_to_test
Failed example:
    sorted(x.name for x in determine_modules_to_test([modules.sql]))
Expected:
    ['examples', 'hive-thriftserver', 'mllib', 'pyspark', 'sparkr', 'sql']
Got:
    ['examples', 'hive-thriftserver', 'mllib', 'pyspark-core', 'pyspark-mllib', 'pyspark-sql', 'pyspark-sql', 'pyspark-streaming', 'sparkr', 'sql']
**********************************************************************
2 items had failures:
   1 of   2 in __main__.determine_modules_for_files
   1 of   3 in __main__.determine_modules_to_test
```

I'll fix this shortly.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: dev/run-tests.py [hunk 2]
```diff
@@ -362,18 +87,21 @@ def determine_modules_to_test(changed_modules):
...
     ['root']
-    >>> sorted(x.name for x in determine_modules_to_test([graphx]))
+    >>> sorted(x.name for x in determine_modules_to_test([modules.graphx]))
     ['examples', 'graphx']
-    >>> sorted(x.name for x in determine_modules_to_test([sql]))
-    ['examples', 'hive-thriftserver', 'mllib', 'pyspark', 'sparkr', 'sql']
+    >>> x = sorted(x.name for x in determine_modules_to_test([modules.sql]))
+    >>> x # doctest: +NORMALIZE_WHITESPACE
+    ['examples', 'hive-thriftserver', 'mllib', 'pyspark-core', \
+     'pyspark-mllib', 'pyspark-sql', 'pyspark-sql', 'pyspark-streaming', \
```
Comment (JoshRosen) [inline dev/run-tests.py]: Whoops, just realized that `pyspark-sql` appears twice; I probably forgot a `set()` call somewhere; will investigate.

File: dev/sparktestsupport/modules.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        "pyspark.mllib.stat.KernelDensity",
+        "pyspark.mllib.tree",
+        "pyspark.mllib.util",
+        "pyspark.mllib.tests",
+    ]
+)
+
+
+pyspark_ml = Module(
+    name="pyspark-sql",
```
Comment (JoshRosen) [inline dev/sparktestsupport/modules.py]: Ah, this should be `pyspark-ml`.

File: dev/sparktestsupport/modules.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    dependencies=[pyspark_core, pyspark_mllib],
+    source_file_regexes=[
+        "python/sql"
```
Comment (JoshRosen) [inline dev/sparktestsupport/modules.py]: This also needs to be updated.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
I also realized that this is going to need a bit of extra work for dev/run-tests to be able to trigger individual Python module tests.  I might end up incorporating some of the ideas from #4269 into this by having `dev/run-tests` invoke `python/run-tests` with a list of module names to test.

**Comment by JoshRosen** [comment]
I've updated this to incorporate passing flags to `python/run-tests` in order to select which suites should be run.

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8583] [SPARK-5482] [BUILD] Refactor python/run-tests to integrate with dev/run-tests module system (#6967)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
It looks like one of the PySpark core tests is failing here:

```python
======================================================================
ERROR: test_termination_sigterm (__main__.DaemonTests)
Ensure that daemon and workers terminate on SIGTERM.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/tests.py", line 1451, in test_termination_sigterm
    self.do_termination_test(lambda daemon: os.kill(daemon.pid, SIGTERM))
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/tests.py", line 1424, in do_termination_test
    daemon = Popen([sys.executable, daemon_path], stdin=PIPE, stdout=PIPE)
  File "/usr/lib64/python2.6/subprocess.py", line 642, in __init__
    errread, errwrite)
  File "/usr/lib64/python2.6/subprocess.py", line 1234, in _execute_child
    raise child_exception
OSError: [Errno 13] Permission denied

======================================================================
ERROR: test_termination_stdin (__main__.DaemonTests)
Ensure that daemon and workers terminate when stdin is closed.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/tests.py", line 1446, in test_termination_stdin
    self.do_termination_test(lambda daemon: daemon.stdin.close())
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/tests.py", line 1424, in do_termination_test
    daemon = Popen([sys.executable, daemon_path], stdin=PIPE, stdout=PIPE)
  File "/usr/lib64/python2.6/subprocess.py", line 642, in __init__
    errread, errwrite)
  File "/usr/lib64/python2.6/subprocess.py", line 1234, in _execute_child
    raise child_exception
OSError: [Errno 13] Permission denied

----------------------------------------------------------------------
Ran 97 tests in 102.677s

FAILED (errors=2)
   Random listing order was used
```

My hunch is that `sys.executable` is returning something different / invalid due our different test invocation strategy (i.e. using a different process structure).

**Comment by JoshRosen** [comment]
This should now be ready for review.  I have another open PR, #7031, which builds on this one to parallelize the Python tests.  Once we review the non-parallel changes here and get this basic version committed, I'll rebase my other PR and we can look at the changes which were necessary to support parallelism.

File: dev/lint-python [hunk 5]
```diff
@@ -19,7 +19,8 @@
 
 SCRIPT_DIR="$( cd "$( dirname "$0" )" && pwd )"
 SPARK_ROOT_DIR="$(dirname "$SCRIPT_DIR")"
-PATHS_TO_CHECK="./python/pyspark/ ./ec2/spark_ec2.py ./examples/src/main/python/"
+PATHS_TO_CHECK="./python/pyspark/ ./ec2/spark_ec2.py ./examples/src/main/python/ ./dev/sparktestsupport"
+PATHS_TO_CHECK="$PATHS_TO_CHECK ./dev/run-tests.py"
```
Comment (davies) [inline dev/lint-python]: python/run-tests.py

File: dev/run-tests.py [hunk 6]
```diff
@@ -504,10 +180,10 @@ def set_title_and_block(title, err_block):
     os.environ["CURRENT_BLOCK"] = ERROR_CODES[err_block]
     line_str = '=' * 72
 
-    print
-    print line_str
-    print title
-    print line_str
+    print()
```
Comment (davies) [inline dev/run-tests.py]: There will be an `()` in Python 2

**Comment by davies** [comment]
LGTM, only two minor issues.

**Comment by JoshRosen** [comment]
@davies, thanks for the review.  I've addressed your two comments, so let's wait and see if this passes tests. Once it passes, I'll merge and rebase my other PRs.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Hopefully that test is just flaky, but let's see.  If it fails again, I'll investigate.

**Comment by JoshRosen** [comment]
Uh oh, looks like the dev/run-tests script has a problem:

```python
Traceback (most recent call last):
  File "./dev/run-tests.py", line 477, in <module>
    main()
  File "./dev/run-tests.py", line 464, in main
    run_python_tests(modules_with_python_tests)
  File "./dev/run-tests.py", line 368, in run_python_tests
    command.append("--modules=%s" % ','.join(m.name for m in modules))
```

I'll investigate and hotfix.

**Comment by JoshRosen** [comment]
Hotfixed in 42db3a1c2fb6db61e01756be7fe88c4110ae638e

**Comment by cocoatomo** [comment]
Hi, @JoshRosen
When running run-tests.py with Python 2.6, I got a folloing error:

```
Running PySpark tests. Output is in python//Users/tomohiko/.jenkins/jobs/pyspark_test/workspace/python/unit-tests.log
Will test against the following Python executables: ['python2.6', 'python3.4', 'pypy']
Will test the following Python modules: ['pyspark-core', 'pyspark-ml', 'pyspark-mllib', 'pyspark-sql', 'pyspark-streaming']
Traceback (most recent call last):
  File "./python/run-tests.py", line 196, in <module>
    main()
  File "./python/run-tests.py", line 159, in main
    python_implementation = subprocess.check_output(
AttributeError: 'module' object has no attribute 'check_output'
```

The cause of this error is using subprocess.check_output function, which exists since Python 2.7.
(ref. https://docs.python.org/2.7/library/subprocess.html#subprocess.check_output)

The paragraph https://spark.apache.org/docs/latest/#downloading says "Spark runs on Java 6+, Python 2.6+...", so should we make run-tests.py enable to run on Python 2.6?

**Comment by JoshRosen** [comment]
Hi @cocoatomo,

If it's not hard to do, I think we should continue to support Python 2.6 for our infra/test scripts.  If you submit a pull request I'll be happy to help review. Hopefully the fix is something simple, such as using the Popen constructor directly instead of using the convenience method.

**Comment by cocoatomo** [comment]
I created the pull request for this: https://github.com/apache/spark/pull/7161.

Thanks.

**Pull request closed without merging**